### PR TITLE
Meta-proposal: ECMA 402 API surface for rust

### PR DESCRIPTION
**Pull request by filmil**
Here is a proposal to make an ECMA 402 style API for rust, which could
admit multiple implementations.

I want to validate the idea before sinking time into doing the actual work.

### CODE REVIEW

**Comment by zbraniecki** [comment]
Stream of thoughts:
* I like that idea!
* I think it should *not* be part of ICU4X, but rather a parallel project.
* `ecma402-rs` should be backed by `rust_icu4c` and as we develop `icu4x`, we should explore migrating it to be supported by `icu4x` instead. We could have it behind a feature first.

File: proposals/pr001-ecma402.md [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+See [README.md](README.md) for the full index of proposals.
+
+
+## Motivation
+
+_Usable text internationalization for Rust, **today**._
+
+### Why now?
+
+* There is a feature gap in Unicode support for rust.
```
Comment (nciric) [inline proposals/pr001-ecma402.md:16]: Nit: Rust

File: proposals/pr001-ecma402.md [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+This doc is submitted for comments as a sanity check on the approach.  Question
+to you, the potential commenter: **do you think this exercise is worth doing?**
+
+
+## Landscape
+
+Rust's support for internationalized text handling is "[still very immature and
+unstable, with most libraries being only weeks old and not used in production
+yet](https://www.arewewebyet.org/topics/i18n/)".  The [ECMA-402 support feature
+request for Rust](https://github.com/rust-lang/rfcs/issues/858) has been open
```
Comment (nciric) [inline proposals/pr001-ecma402.md:41]: I've replied to this issue and pointed them to icu4x repo. Thanks for finding it (2018 is the last time it was updated).

File: proposals/pr001-ecma402.md [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+only in Fuchsia OS, it **makes it cheap** to contribute internationalized text
+support thanks to ICU providing the heavy lifting.  To illustrate: it took me a
+morning and half an afternoon to provide [collation (locale-sensitive
+sorting)](https://crates.io/crates/rust_icu_ucol) support for rust building on
+top of `rust_icu`.  There were no rust collation libraries prior to that
+afternoon.
+
+In parallel, ICU4X is a project that will provide next-generation support for
+internationalized text, based on a Rust implementation of Unicode standards.
+This project is still some time away from first usable functionality, is my
```
Comment (nciric) [inline proposals/pr001-ecma402.md:81]: I would make whole proposal less personal and more fact based:

This project is still in early phase of development and not ready for general use. (no need for my estimate, first usable functionality etc).

File: proposals/pr001-ecma402.md [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+existed; but we are unwilling to contribute because it's expensive to make.
+The status quo continues.
+
+I'll present an assumption: **if robust text internationalization support were
+available for rust today, then people who need it would use it in their rust
+programs**.  I tested this assumption with
+[rust\_icu](https://github.com/google/rust_icu), a rust wrapper for the
+[Unicode's ICU library](http://site.icu-project.org/home).  While the jury is
+still out as to where this goes, given that the library is still [relatively
+unknown](https://www.google.com/search?q=rust+unicode+support), and used mostly
```
Comment (nciric) [inline proposals/pr001-ecma402.md:71]: Maybe cut this part (no additional value, more words). Just say rust_icu is not widely adopted except in Fuchsia OS.

File: proposals/pr001-ecma402.md [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+1.  Implement the ECMA 402 API from (1) with `rust_icu`.
+    a.  (Welcome any alternative implementations as well!)
+1.  Publish the implementation (2) for use.
+1.  Once ICU4X is ready, implement the API from (1) with ICU4X.
+	a.  This does not preclude a "transitional" implementation where some
+	functionality is partially provided by `rust_icu`, and partially by ICU4X.
+	It may not be very efficient in that way, but allows for a smooth
+	transition into the future.
+
+**Why ECMA 402 (versus something else)?** ECMA 402 a well understood API.  We
```
Comment (nciric) [inline proposals/pr001-ecma402.md:107]: 402 *is* a well understood

File: proposals/pr001-ecma402.md [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+
+## Requirements
+
+*   **Conformance.**
+
+	The API should conform to the ECMA 402 specification.
+
+*   **Generality.**
+
+	It should be possible to implement the API with rust\_icu, as well as with
```
Comment (nciric) [inline proposals/pr001-ecma402.md:127]: remove \

File: proposals/pr001-ecma402.md (hunk 5 above)
Comment (filmil) [reply proposals/pr001-ecma402.md]: Done.

File: proposals/pr001-ecma402.md (hunk 6 above)
Comment (filmil) [reply proposals/pr001-ecma402.md]: Done

**Comment by filmil** [comment]
> Stream of thoughts:

Thank you!

> * I like that idea!

Ok, that seems like a good start. :)

> * I think it should _not_ be part of ICU4X, but rather a parallel project.

Would you be OK with at least the trait definitions having at least a temporary home here?  I think that could be easier to start, and keeps the proposals close to where the discussion happens.

> * `ecma402-rs` should be backed by `rust_icu4c` and as we develop `icu4x`, we should explore migrating it to be supported by `icu4x` instead. We could have it behind a feature first.

SGTM.

**Comment by zbraniecki** [comment]
> Would you be OK with at least the trait definitions having at least a temporary home here? I think that could be easier to start, and keeps the proposals close to where the discussion happens.

Hmm, I wouldn't block that, but I'm a bit confused - maybe I don't fully grasp your vision for that.

ICU4X is solidifying as a project that produces a set of libraries and in the future will produce outputs (say, for WASM etc.).
What is the benefit of using this repository to store trait definitions for another project/repository? If you'll store them on `master`, you're adding code that feels experimental and belonging to a branch or separate repo, and if you're adding it on a branch, then what's the benefit over a separate repo?

I understand the value of using the ICU4X organization and brainpower to review/validate the traits, but I hope that we can file an issue here and discuss/review code in a another repo.

If it's just that you prefer to not move away from icu4x, I'd at least encourage forking icu4x and using a branch to store it.

I may be also just misunderstanding how you see the relationship between those traits and ICU4X.

**Comment by sffc** [comment]
Sounds like a good meeting agenda item.

**Comment by filmil** [comment]
> I understand the value of using the ICU4X organization and brainpower to review/validate the traits, ...

Yes indeed, the main thing I'd like to have is the ability to get feedback from the group in a timely manner.  Where exactly the code lives is secondary; some red tape may be involved in spinning up a new repo, but that's doable.

**Comment by filmil** [comment]
Would like to get to an approval and merge for this proposal.  I can haz?

File: proposals/pr001-ecma402.md [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+[fmil@google.com](mailto:fmil@google.com)
+
+2020-05-21
+
+See [README.md](README.md) for the full index of proposals.
+
+
+## Motivation
+
+_Usable text internationalization for Rust, **today**._
```
[the remark this answers is not in the document]
Comment (filmil) [reply proposals/pr001-ecma402.md:12]: I view it as a strict subset.  Compare to internationalizing calendars, cultural references, pictorial representations (icons) etc.

[the remark this answers is not in the document]
Comment (zbraniecki) [reply proposals/pr001-ecma402.md:12]: I see. Makes sense. 
It's not a specialization term I've seen in relation to ECMA402 which describes itself as an Internationalization API. It's a nit, and I'm comfortable with your call.

File: proposals/pr001-ecma402.md [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+
+### Why now?
```
[the remark this answers is not in the document]
Comment (filmil) [reply proposals/pr001-ecma402.md:14]: Done

**Comment by filmil** [comment]
Done with the immediately actionable changes.  PTAL?

**Comment by zbraniecki** [comment]
The final document looks very good. Thank you for taking time to apply my non-blocking feedback!

**Pull request merged**